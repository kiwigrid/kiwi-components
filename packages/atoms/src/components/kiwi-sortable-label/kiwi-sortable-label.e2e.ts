import { newE2EPage } from '@stencil/core/testing';

describe('kiwi-sortable-label', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<kiwi-sortable-label></kiwi-sortable-label>');

    const element = await page.find('kiwi-sortable-label');
    expect(element).toHaveClass('hydrated');
  });
});
