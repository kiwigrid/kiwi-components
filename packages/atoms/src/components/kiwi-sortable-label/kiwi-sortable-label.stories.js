import { html } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/SortableLabel',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

export const basic = (args) => {
  const { label, sortDirection } = args;

  return html`
    <div class="panel panel-default">
      <div class="panel-body">
        <kiwi-sortable-label label=${label} sort-direction=${sortDirection} />
      </div>
    </div>
  `;
};
basic.args = {
  label: 'Label',
};
basic.argTypes = {
  sortDirection: {
    control: {
      type: 'select',
      options: ['asc', 'desc', 'off'],
    },
    defaultValue: 'asc',
  },
};
