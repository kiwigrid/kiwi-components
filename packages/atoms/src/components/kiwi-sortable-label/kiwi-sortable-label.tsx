import {
  Component,
  ComponentInterface,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
} from '@stencil/core';

export type SortDirection = 'asc' | 'desc' | 'off';

@Component({
  tag: 'kiwi-sortable-label',
  styleUrl: 'kiwi-sortable-label.css',
  shadow: false,
})
export class KiwiSortableLabel implements ComponentInterface {
  /**
   * The label to be displayed.
   */
  @Prop()
  label?: string;

  /**
   * The key to be sent by sort event.
   */
  @Prop()
  sortKey!: string;

  /**
   * The initial sort direction.
   */
  @Prop()
  sortDirection: SortDirection = 'off';

  /**
   * Sort event sent when clicked on a label.
   */
  @Event()
  sort!: EventEmitter<string>;

  render(): JSX.Element {
    const sortIcon =
      this.sortDirection === 'desc'
        ? 'glyphicon-sort-by-alphabet-alt'
        : 'glyphicon-sort-by-alphabet';

    return (
      <div class="text-nowrap">
        <a class="header pointer" onClick={this.handleClick}>
          {this.label == undefined ? <slot /> : this.label}
          <span class={`glyphicon ${sortIcon}`}></span>
        </a>
      </div>
    );
  }
  private handleClick = (): void => {
    this.sort.emit(this.sortKey);
  };
}
