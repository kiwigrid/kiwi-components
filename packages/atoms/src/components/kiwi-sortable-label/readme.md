# kiwi-sortable-label



<!-- Auto Generated Below -->


## Properties

| Property               | Attribute        | Description                       | Type                       | Default     |
| ---------------------- | ---------------- | --------------------------------- | -------------------------- | ----------- |
| `label`                | `label`          | The label to be displayed.        | `string \| undefined`      | `undefined` |
| `sortDirection`        | `sort-direction` | The initial sort direction.       | `"asc" \| "desc" \| "off"` | `'off'`     |
| `sortKey` _(required)_ | `sort-key`       | The key to be sent by sort event. | `string`                   | `undefined` |


## Events

| Event  | Description                              | Type                  |
| ------ | ---------------------------------------- | --------------------- |
| `sort` | Sort event sent when clicked on a label. | `CustomEvent<string>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
