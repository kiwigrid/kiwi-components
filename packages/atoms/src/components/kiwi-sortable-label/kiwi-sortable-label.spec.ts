import { newSpecPage } from '@stencil/core/testing';
import { KiwiSortableLabel } from './kiwi-sortable-label';

describe('kiwi-sortable-label', () => {
  it('builds', () => {
    expect(new KiwiSortableLabel()).toBeTruthy();
  });

  it('renders', async () => {
    const label = await newSpecPage({
      components: [KiwiSortableLabel],
      html: `
        <kiwi-sortable-label
          label="Random"
          criterion="math.random"
          sort-key="random"
        ></kiwi-sortable-label>`,
    });

    // cycle through each sort state and finally click remove sorting

    expect(label.root).toMatchSnapshot('initial');
  });

  it('emits sort event', async () => {
    const label = await newSpecPage({
      components: [KiwiSortableLabel],
      html: `
        <kiwi-sortable-label
          label="GrÃ¼n"
          sort-key="gruen"
        ></kiwi-sortable-label>`,
    });

    const sortSpy = jest.fn();
    label.root!.addEventListener('sort', sortSpy);

    const a = label.root!.querySelector('a');
    a!.click();
    expect(sortSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'sort',
        detail: 'gruen',
      }),
    );
  });
});
