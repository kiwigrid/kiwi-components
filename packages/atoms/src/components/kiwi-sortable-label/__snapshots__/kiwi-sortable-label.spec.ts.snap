// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-sortable-label renders: initial 1`] = `
<kiwi-sortable-label criterion="math.random" label="Random" sort-key="random">
  <!---->
  <div class="text-nowrap">
    <a class="header pointer">
      Random
      <span class="glyphicon glyphicon-sort-by-alphabet"></span>
    </a>
  </div>
</kiwi-sortable-label>
`;
